cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,216 @@
+                            column_view const& right_edges,
+                            rmm::mr::device_memory_resource* mr)
+{
+  auto output = cudf::make_numeric_column(data_type(type_to_id<size_type>()), input.size());
+  auto output_mutable_view = output->mutable_view();
+  auto input_device_view   = column_device_view::create(input);
+
+  thrust::transform(thrust::device,
+                    input_device_view->pair_begin<T, InputIsNullable>(),
+                    input_device_view->pair_end<T, InputIsNullable>(),
+                    output_mutable_view.begin<size_type>(),
+                    // Must specify const T as the template type because the column
+                    // views provided on the edges will always return const types. The
+                    // template arguments to `datq` need not specify const since that
+                    // const is added as part of the signature.
+                    bin_finder<const T, LeftComparator, RightComparator>(
+                      detail::device_span<const T>(left_edges.data<T>(), left_edges.size()),
+                      detail::device_span<const T>(right_edges.data<T>(), right_edges.size())));
+
+  auto mask_and_count = cudf::detail::valid_if(output_mutable_view.begin<size_type>(),
+                                               output_mutable_view.end<size_type>(),
+                                               filter_null_sentinel());
+
+  output->set_null_mask(mask_and_count.first, mask_and_count.second);
+  return output;
+}
+
+template <typename T>
+constexpr inline auto is_supported_bin_type()
+{
+  // TODO: Determine what other types (such as fixed point numbers) should be
+  // supported, and whether any of them (like strings) require special
+  // handling.
+  return (cudf::is_numeric<T>() && !std::is_same<T, bool>::value);  // || cudf::is_fixed_point<T>();
+}
+
+}  // anonymous namespace
+}  // namespace detail
+
+/// Functor suitable for use with type_dispatcher that exploits SFINAE to call the appropriate
+/// detail::bin method.
+struct bin_type_dispatcher {
+  template <typename T, typename... Args>
+  std::enable_if_t<not detail::is_supported_bin_type<T>(), std::unique_ptr<column>> operator()(
+    Args&&... args)
+  {
+    CUDF_FAIL("Type not support for cudf::bin");
+  }
+
+  template <typename T>
+  std::enable_if_t<detail::is_supported_bin_type<T>(), std::unique_ptr<column>> operator()(
+    column_view const& input,
+    column_view const& left_edges,
+    inclusive left_inclusive,
+    column_view const& right_edges,
+    inclusive right_inclusive,
+    rmm::mr::device_memory_resource* mr)
+  {
+    // Note: We could be slightly more efficient in the not nullable case
+    // by overloading the call operator of bin_finder to accept a value in
+    // addition to a pair and using a raw (non-pair) iterator in the
+    // transform call in detail::bin, if we need to make this faster.
+    if (input.nullable()) {
```

Comment:
Be careful with compile time here. This is a lot of kernel instantiations.

Reply:
That's true, I wasn't sure how we should prioritize this. The alternative is dispatching more of these at run time, what would you suggest?

Reply:
I suspect putting the nullable/null check inside the kernel wouldn't have much impact on performance.

Reply:
I agree, that's fine with me then.

Regarding the check, in general in cudf do we tend to make decisions based on `input.nullable()` or `input.null_count() == 0`? This question is also relevant [here](https://github.com/rapidsai/cudf/pull/7554/files#diff-637d33f0cb45bee09388ad24ce306f6fb277d8ea8acae5bdb8e29d68f4833cf4R198-R202).